### PR TITLE
Guards, more defensive VO and IDs

### PR DESCRIPTION
**Pull request by EnricoTirotta**
## Current Request - _DDD Model_

I create **Guards.cs** for adding some Assertion about _"general"_ correct state of an object in a specific moment. It's a good utility inside constructors and methods. You can change all _"base" if-checkers_ with something more _meaningful_. From my point of view, it's a more clear way to read model.
 
In Address Value Object, i try a more defensive approch with guards. I **remove getters** because i think we should add in future, if needed. I add a static constructor **_fromString_** that take an address string in input. 
Now, we may also discuss about: 
- correct string format in Address VO (see: Address toString() and inside static constructor fromString: { .... string[] addressElements = address.Split('_');  ... })
- What can be null in an address?
- Where the lower bounds are in each string field?

Leave  _private_  accessibility is a choice, in model design i love explicit things.

In OrderAggregateTest i write some simple _[Theory]_ with multiple input for checking Address VO guards and behaviour (static constructor).

This is a proposal to know what you think about. Do you think I have a **too much defensive** approach? 

## Possible future modification about _ids_  - _DDD Model_
We may add all **IDs** inside a Value Object. For example: PaymentMethodId _paymentId; BuyerId _buyerId;

I think it should be a good update for DDD project: 
1. **Guards** before assign the id (i know, i dream guards at night 😄 )
2. Less error prone. **Compiler can help us!** Assign an **int** has no meaning, it's an int. So if we assign a _paymentMethodId(int) inside a _buyerId(int), compiler says us "it's all ok guys!! 🎉 " ... 
3. We protect every piece of code that uses these Ids. The changes happen, with this approach we can change Id types without impact code.

Have a nice code day,
Enrico

### CODE REVIEW

**Comment by ardalis** [comment]
You might like this package:
https://github.com/ardalis/GuardClauses

**Comment by EnricoTirotta** [comment]
Thanks @ardalis , i did not know this project. Assertions are a great usefull tool. I will give an accurate read next days.

File: src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+        {
+            Guards.AssertStringIsNotEmptyOrNull(address);
+            Guards.AssertStringLowerBound(address,5);
+
+            //Address toString() format with '_' between fields
+            string[] addressElements = address.Split('_');
+            Guards.AssertIsTrue(addressElements.Length == 5, "Invalid Address string format");
+
+            //toString() order
+            String Street = addressElements[0];
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs]: Local variables should be in camelCase, not PascalCase.

File: src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs [hunk 2]
```diff
@@ -1,26 +1,58 @@
...
 {
-    public class Address : ValueObject
+    public sealed class Address : ValueObject
     {
-        public String Street { get; private set; }
-        public String City { get; private set; }
-        public String State { get; private set; }
-        public String Country { get; private set; }
-        public String ZipCode { get; private set; }
+        private String Street;
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs]: Private fields should be in camelCase or _camelCase, not PascalCase.

File: src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs [hunk 3]
```diff
@@ -1,26 +1,58 @@
...
+        private String City;
+        private String State;
+        private String Country;
+        private String ZipCode;
 
-        private Address() { }
+        public static Address fromString(String address)
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/AggregatesModel/OrderAggregate/Address.cs]: Methods should be in PascalCase and not in camelCase

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                throw new InvalidOperationException(message);
+            }
+        }
+
+        public static void AssertIntIsGreaterThan0(int number)
+        {
+            if (number < 0)
+            {
+                string message = $"Invalid int value. Set an interger greater than 0";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentOutOfRangeException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        }
+
+        public static void AssertStringRange(string arg1, int upperBound, int lowerBound)
+        {
+            AssertNotNull(arg1);
+            if (ArgumentCharCount(arg1) > upperBound || ArgumentCharCount(arg1) < lowerBound)
+            {
+                string message = $"Invalid string range. Lower bound rule:{lowerBound} and Upper bound rule:{upperBound}. Whitespace borders are not allowed";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentOutOfRangeException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+        }
+
+        public static void AssertStringIsNotEmptyOrNull(string arg)
+        {
+            if (String.IsNullOrEmpty(arg))
+            {
+                string message = $"Invalid string value";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentNullException or ArgumentOutOfRangeException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+        }
+
+        public static void AssertStringUpperBound(string arg1, int upperBound)
+        {
+            AssertNotNull(arg1);
+            if (ArgumentCharCount(arg1) > upperBound)
+            {
+                string message = $"Invalid string upper bound. Upper bound rule:{upperBound}. Whitespace borders are not allowed";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentOutOfRangeException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+        }
+
+        public static void AssertStringLowerBound(string arg1, int lowerBound)
+        {
+            AssertNotNull(arg1);
+            if (ArgumentCharCount(arg1) < lowerBound)
+            {
+                string message = $"Invalid string lower bound. Lower bound rule:{lowerBound}. Whitespace borders are not allowed";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentOutOfRangeException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+        }
+
+        public static void AssertIsFalse(bool flag, string message)
+        {
+            if (flag)
+            {
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+        }
+
+        public static void AssertIsTrue(bool flag,string message)
+        {
+            if (!flag)
+            {
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        public static void AssertGuidsAreDifferent(Guid Id1, Guid Id2)
+        {
+            AssertGuidNotEmpty(Id1);
+            AssertGuidNotEmpty(Id2);
+            if (Id1 == Id2)
+            {
+                string message = $"Invalid request, the GUIDs are equal. Guid 1 {Id1} and Guid 2 {Id2}";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentException

File: src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+        }
+
+        public static void AssertGuidNotEmpty(Guid Id)
+        {
+            if (Id == Guid.Empty)
+            {
+                string message = $"Invalid GUID value. Set an ID not empty";
+                throw new InvalidOperationException(message);
```
Comment (CamiloTerevinto) [inline src/Services/Ordering/Ordering.Domain/SeedWork/Guards.cs]: This should throw ArgumentNullException

**Comment by mvelosop** [comment]
Hi @EnricoTirotta, I'm sorry, but I'm closing this PR now, because the changes here break code that reference the Address VO, and the properties need to be public for the correct Order serialization and private set for deserialization.

Also, the validations added here are redundant with the `CreateOrderCommandValidator`.

Regards.

**Pull request closed without merging**